### PR TITLE
[Clarity2] Convert buffer to uint and int

### PR DESCRIPTION
**Pull request by gregorycoppola**
## Description
Fixes 2616

This PR adds Clarity functions to convert a 16-byte buffer to 1) int and 2) uint, using encodings 1) big-endian and 2) little-endian.

The functions are guarded to require ClarityVersion::Clarity2.

## Testing
I tested this by creating and running the new tests in src/vm/tests/conversions.rs

## Here are some issues to consider
- [ ] To make this work, I added the function 'BuffData::as_slice'.. is that a good idea?
- [ ] I added new descriptions to the vm::docs, but didn't add any examples because when I tried I got the error NonFunctionApplication. Should we be able to add new functions and new docs at the same time?
- [ ] Should we allow < 16 bytes?

### CODE REVIEW

File: src/chainstate/stacks/boot/costs.clar
Comment (kantai) [inline src/chainstate/stacks/boot/costs.clar:165]: Unfortunately we can't make cost-contract changes this way -- with this change, the "costs" contract published at genesis would change, which would change the consensus value for the stacks blockchain starting at height = 0. This is basically issue #2682. For now, I'd recommend just reusing an existing cost function for the new methods, and leaving a todo comment.

File: src/vm/analysis/type_checker/natives/conversions.rs
Comment (kantai) [inline src/vm/analysis/type_checker/natives/conversions.rs:36]: For these functions, the type checker should do an admission check, rather than an exact type check.

e.g.,

`(buff-to-uint 0x00)` should work, because we'll do left-padding, but the type of that input is `(buff 1)`, which would fail the exact type check.

Instead, you can do something like:

`checker.type_check_expects(&args[0], context, &TypeSignature::SequenceType(SequenceSubType::BufferType(16)))`

Rather than the current match check.

The same applies to `buff-to-uint`.

File: src/vm/docs/mod.rs
Comment (kantai) [inline src/vm/docs/mod.rs:497]: It'd be good to include some examples for these functions -- they'd double as tests as well.

File: src/vm/functions/conversions.rs
Comment (kantai) [inline src/vm/functions/conversions.rs:73]: Do these methods need to be "special"? I think they could be implemented as `CallableType::NativeFunction`, because they don't need to do any special type checks (normal type checks work fine) and they don't need any special access to the `env` or `context`.

File: src/chainstate/stacks/boot/costs.clar
Comment (gregorycoppola) [reply src/chainstate/stacks/boot/costs.clar]: I reverted this now. The cost should now be specified in src/vm/functions/mod.rs, ClarityCostFunction::Mul.

File: src/vm/analysis/type_checker/natives/conversions.rs
Comment (gregorycoppola) [reply src/vm/analysis/type_checker/natives/conversions.rs]: I reverted this file. Checking should now be done in src/vm/analysis/type_checker/natives/mod.rs

Is there a way to test the type checking?

File: src/vm/functions/conversions.rs
Comment (gregorycoppola) [reply src/vm/functions/conversions.rs]: These should be native functions now.. see src/vm/analysis/type_checker/natives/mod.rs. I also changed the function signatures to not use the 'env' and 'context'.

File: src/vm/docs/mod.rs
Comment (gregorycoppola) [reply src/vm/docs/mod.rs:497]: When I try to add the examples, they fail with the error

`err: NonFunctionApplication, expressions: Some([SymbolicExpression { expr: Atom(ClarityName("buff-to-int-le"))...`

Is this a clarity version issue?

File: src/vm/analysis/type_checker/natives/mod.rs
Comment (gregorycoppola) [inline src/vm/analysis/type_checker/natives/mod.rs:575]: Is this right? Is there a good way to test it?

File: src/vm/tests/costs.rs
Comment (gregorycoppola) [inline src/vm/tests/costs.rs:85]: Should any more thought be given to this test?

**Comment by gregorycoppola** [comment]
I updated the code with 1) answers to the comments, and 2) logic to support < 16 byte input.

The only thing that wasn't addressed is the question of the docs, because it doesn't seem to work when I put in examples for these new functions. Is it a clarity version isssue? see https://github.com/blockstack/stacks-blockchain/pull/2691#discussion_r647003318

File: src/vm/analysis/errors.rs
Comment (gregorycoppola) [inline src/vm/analysis/errors.rs:134]: Probably I should use an existing error like "TypeError(TypeSignature, TypeSignature)" instead of this.

File: src/vm/analysis/type_checker/natives/conversions.rs
Comment (kantai) [reply src/vm/analysis/type_checker/natives/conversions.rs]: Yep -- look at the unit tests in `vm/analysis/type_checker/tests/mod.rs` -- you should be able to add some tests for different input types pretty easily, and test them against either expected errors or expected result types.

File: src/vm/docs/mod.rs
Comment (kantai) [reply src/vm/docs/mod.rs]: There shouldn't be a versioning issue -- what's the example that you're trying to add?

File: src/vm/tests/costs.rs
Comment (kantai) [reply src/vm/tests/costs.rs:85]: This is fine for now -- we'll want more precise testing once the benchmarking work progresses though.

File: src/vm/analysis/errors.rs
Comment (kantai) [reply src/vm/analysis/errors.rs]: Yes -- if you use `checker.type_check_expects(..)` to perform the type checks, then it will create the `TypeError(..)` error for you, so you should be able to remove the `ExpectedBuffer16` variant.

Comment (gregorycoppola) [reply src/vm/analysis/errors.rs]: The error in 'src/vm/functions/conversions.rs' is now:

                return Err(CheckErrors::TypeError(
                    SequenceType(BufferType(BufferLength(16))),
                    TypeSignature::type_of(&value),
                )

File: src/vm/analysis/type_checker/natives/conversions.rs
Comment (gregorycoppola) [reply src/vm/analysis/type_checker/natives/conversions.rs]: I added some tests in `vm/analysis/type_checker/tests/mod.rs`

File: src/vm/docs/mod.rs
Comment (gregorycoppola) [reply src/vm/docs/mod.rs]: It was a version issue.. there was a Clarity1 assumption in the test.. I changed it and added a bunch of examples.

**Comment by gregorycoppola** [comment]
Thanks for the notes. I added some fixes.

File: src/vm/docs/mod.rs
Comment (gregorycoppola) [inline src/vm/docs/mod.rs:1941]: note.. Clarity version referenced here.

**Comment by gregorycoppola** [comment]
@kantai Hopefully most major issues are fixed now.

File: src/vm/analysis/arithmetic_checker/mod.rs
Comment (kantai) [inline src/vm/analysis/arithmetic_checker/mod.rs:196]: I think that these buff conversion functions should not pass the arithmetic only checker -- so these new functions would be "FunctionNotPermitted".

File: src/vm/functions/mod.rs
Comment (kantai) [inline src/vm/functions/mod.rs:278]: Can you copy this TODO comment to the other instances, too? It'll help us make sure we don't accidentally forget to update one.

File: src/vm/functions/conversions.rs
Comment (kantai) [inline src/vm/functions/conversions.rs:54]: You can use the `TypeValue` error to throw a CheckError without doing `type_of`.

File: src/vm/docs/mod.rs
Comment (kantai) [inline src/vm/docs/mod.rs:179]: Can you add a description of the padding behavior to the descriptions of these functions?

**Review by kantai — COMMENTED** [review]
This looks good to me -- just some slight tweaks and a request for some more description in the docs.

File: src/vm/analysis/arithmetic_checker/mod.rs
Comment (gregorycoppola) [reply src/vm/analysis/arithmetic_checker/mod.rs:196]: done

File: src/vm/docs/mod.rs
Comment (gregorycoppola) [reply src/vm/docs/mod.rs:179]: done

File: src/vm/functions/mod.rs
Comment (gregorycoppola) [reply src/vm/functions/mod.rs:278]: done

File: src/vm/functions/conversions.rs
Comment (gregorycoppola) [reply src/vm/functions/conversions.rs]: is this what you're thinking?

**Comment by gregorycoppola** [comment]
> This looks good to me -- just some slight tweaks and a request for some more description in the docs.

Thanks.. addressed the comments. Not sure if the new type checking is right? I made the s/TypeError/TypeValueError/ change in the production code, but left the test unchanged. The same tests are still passing.

File: src/vm/functions/mod.rs
Comment (pavitthrap) [inline src/vm/functions/mod.rs:278]: just want to double check if we want to merge these "todos" as well

File: src/vm/mod.rs
Comment (pavitthrap) [inline src/vm/mod.rs:358]: Do we still want this comment above `execute`

**Review by pavitthrap — APPROVED** [review]
lgtm!

File: src/vm/docs/mod.rs
Comment (jcnelson) [inline src/vm/docs/mod.rs:176]: Do we want to annotate the documentation somehow to indicate that this is only available in Clarity2?  Doesn't have to be in this PR, but if not, can you make a separate issue so we remember to do this?

File: src/vm/docs/mod.rs
Comment (jcnelson) [inline src/vm/docs/mod.rs:217]: Typo -- `buff-to-int-be`

File: src/vm/docs/mod.rs
Comment (jcnelson) [inline src/vm/docs/mod.rs:232]: Typo -- `buff-to-uint-le`

File: src/vm/functions/conversions.rs
Comment (jcnelson) [inline src/vm/functions/conversions.rs:57]: Minor style nit: this could be `let mut buf = [0u8; 16]`.

File: src/vm/functions/conversions.rs
Comment (jcnelson) [inline src/vm/functions/conversions.rs:29]: Do we need a specific `EndianDirection` type?  Could we get away with instead defining four different `conversion_fn` functions that simply wrapped `::from_le_bytes` and `::from_be_bytes` for `u128` and `i128`?

File: src/vm/functions/conversions.rs
Comment (jcnelson) [inline src/vm/functions/conversions.rs:58]: You might be able to get away with doing an internal `malloc()` here if you instead copied the original slice directly into `buf` (no need to reverse it), and passed _that_ into the relevant `::from_be_bytes()` or `::to_be_bytes()` methods.

File: src/vm/functions/mod.rs
Comment (jcnelson) [reply src/vm/functions/mod.rs:278]: I don't think there's yet a way to represent cost functions for Clarity2.  If not, then @gregorycoppola please create issues to do so, and tag them as `stacks-2.1`.  Thanks!

File: src/vm/docs/mod.rs
Comment (gregorycoppola) [reply src/vm/docs/mod.rs:176]: see https://github.com/blockstack/stacks-blockchain/issues/2716

I also added some text for now:
`    Note: This function is only available starting with Stacks 2.1.`

File: src/vm/functions/mod.rs
Comment (gregorycoppola) [reply src/vm/functions/mod.rs:278]: @pavitthrap I had previously merged the todos but @kantai asked to expand them.
@jcnelson I added https://github.com/blockstack/stacks-blockchain/issues/2717

File: src/vm/functions/conversions.rs
Comment (gregorycoppola) [reply src/vm/functions/conversions.rs:29]: The endian direction has a use apart from this, because we need to know which side to do the padding on.

File: src/vm/functions/conversions.rs
Comment (gregorycoppola) [reply src/vm/functions/conversions.rs:58]: I made some changes to maybe make it easier to understand. And, we avoid the 'reverse'.

File: src/vm/mod.rs
Comment (kantai) [reply src/vm/mod.rs:358]: Oh, yes, we still want this comment -- should be changed to `Only used by CLI and unit tests`

Comment (gregorycoppola) [reply src/vm/mod.rs:358]: done

**Review by jcnelson — APPROVED** [review]
LGTM! Thanks @gregorycoppola

**Review by kantai — APPROVED** [review]
This LGTM! For merging, please wait until #2659 merges, and then you can merge this directly to `next`

**Pull request merged**
